File: leetcode/515/solution.go
package main

import "math"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func largestValues(root *TreeNode) []int {
	var answer []int
	queue := []*TreeNode{}

	if root != nil {
		queue = append(queue, root)
	}

	i := -1
	for len(queue) > 0 {
		l := len(queue)
		answer = append(answer, math.MinInt32)
		i += 1
		for j := 0; j < l; j++ {
			node := queue[0]
			queue = queue[1:]
			if answer[i] < node.Val {
				answer[i] = node.Val
			}
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
	}

	return answer
}

func main() {
	largestValues(&TreeNode{})
}

File: leetcode/515/solution.js
/**
 * https://leetcode.com/problems/find-largest-value-in-each-tree-row/
 * 515. Find Largest Value in Each Tree Row
 */

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var largestValues = function (root) {
  const answer = [];
  const queue = [root];
  const MAX = (Math.pow(2, 31) + 1) * -1;
  if (!root) return [];

  while (queue.length) {
    const l = queue.length;
    let maxValue = MAX;
    for (let i = 0; i < l; i++) {
      const node = queue.shift();
      maxValue = Math.max(maxValue, node.val);
      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);
    }
    answer.push(maxValue);
  }
  return answer;
};
